site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,262 @@
+
+Configure the following attributes and then save the new user:
+* **Email**
+* **Password and Confirm**: Toggle `Send email to setup password` to off and manually enter a password for the user.
+
+### Register the User
+Looking good! Now we have our new application and a user. The last thing we want to do is `register` the user with our `application`.
+Again, navigate to the `Users` page, find our newly created user, and click the `Manage` button.
+
+Click on the `Add Registration` button, select the Rails application we created, and save.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/user-registered.png" alt="User Registered" class="img-fluid" figure=false %}
+
+## Build the Rails Application
+Lets get started building a very basic Rails application to demonstrate authentication with FusionAuth.
+
+We are going to start by creating a new Rails application.
+```bash
+rails new rails-fusionauth-app
+```
+
+
+
+### Configuration
+Before we go any further, there are some critical `Application` attributes we will need from our FusionAuth configuration so that our Rails app knows how to successfully communicate with it.
+Specifically:
+* `client_id`
+* `client_secret`
+* `Authorized redirect URL's`
+* `Logout URL`
+
+You can recall, these can be found under the `OAuth` tab when modifying an application in FusionAuth.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/edit-application.png" alt="Edit Application" class="img-fluid" figure=false %}
+
+### Routes
+For this app, we are going to setup three routes:
+* **welcome page**: Our root page or `view`. 
+* **oauth2-callback** endpoint: Receiver of the redirect after logging into FusionAuth.
+* **logout** endpoint: Receiver of the redirect after logging out of FusionAuth.
+
+```ruby
+Rails.application.routes.draw do
+  root to: 'welcome#index'
+  get '/oauth2-callback', to: 'oauth#oauth_callback'
+  get '/logout', to: 'oauth#destroy'
+end
+```
+
+### Authenticating and Logging Out 
+We will start off with a very basic welcome view that contains a link for a user to login to the application.
+
+```ruby
+# app/views/welcome/index.html.erb
+
+<p>Welcome user! Please <%= link_to 'Log In', @login_url %></p>
+```
+
+To begin the login process, we will want to construct a URL to make an [Authorization Code Grant Request](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#authorization-code-grant-request).
+This URL is composed of the `client_id`, `client_secret`, and `redirect_uri` values from our application configuration. 
+
+We will also construct the URL used to make a [Logout](https://fusionauth.io/docs/v1/tech/oauth/endpoints/#logout) request.
+
+```ruby
+# app/controllers/welcome_controller.rb
+
+class WelcomeController < ApplicationController
+  def initialize
+    @login_url = 'http://localhost:9011/oauth2/authorize?client_id=799b3fe1-4a52-494f-a13a-b97b1377e073&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Foauth2-callback'
```

Comment:
Is there a rails gem to do this? Something analogous to https://github.com/thephpleague/oauth2-client

https://github.com/oauth-xx/oauth2 looks promising. 

I know a lot of our example apps build this url directly, but I think at least a nod to the library is helpful. Happy to discuss more.